### PR TITLE
feat: prefer scope for data produced by middlewares

### LINKED ISSUE #73: Share middleware data via its Scope
To improve test comprehensibility and general app maintainability, store values produced from middlewares such as the `ctxvar_geolocation` value, found in `middleware/geolocation.py`, to the `scope` dictionary, an injected dependency.

!image-20221013-220930.png|width=697,height=546!

**Related documentation:**

[https://github.com/encode/starlette/blob/master/docs/middleware.md#passing-information-to-endpoints](https://github.com/encode/starlette/blob/master/docs/middleware.md#passing-information-to-endpoints|smart-link)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/CONSVC-2021)
┆Attachments: <a href="https://mozilla-hub.atlassian.net/rest/api/2/attachment/content/29696">image-20221013-220930.png</a>
┆epic: Merino: maintenance

### PR DESCRIPTION
**Pull request by pjenvey**
Closes #73

### CODE REVIEW

File: merino/middleware/logging.py [hunk 1]
```diff
@@ -38,8 +35,8 @@ async def send_wrapper(message: Message) -> None:
             if message["type"] == "http.response.start":
                 request = Request(scope=scope)
                 if PATTERN.match(request.url.path):
-                    location = ctxvar_geolocation.get()
-                    ua = ctxvar_user_agent.get()
+                    location = scope["merino_geolocation"]
+                    ua = scope["merino_user_agent"]
```
Comment (Trinaa) [inline merino/middleware/logging.py:39]: What do you think of capturing the `merino_geolocation` and `merino_user_agent` string keys in a ScopeKey Enum or Constant, so as to prevent errors from misspelling or facilitate future re-use/refactor?

File: tests/unit/providers/test_accuweather.py [hunk 2]
```diff
@@ -244,65 +252,59 @@ async def test_invalid_location_key(accuweather: Provider) -> None:
...
+        city=None,
+        dma=None,
+        postal_code=None,
     )
 
     set_response_bodies()
 
-    res = await accuweather.query("")
+    res = await accuweather.query(SuggestionRequest("", geolocation))
     assert res == []
```
Comment (Trinaa) [inline tests/unit/providers/test_accuweather.py:310]: This is a bit out of scope, so let me know if you prefer I make a separate PR.

I realized looking at this changeset, that these last 3 tests (`test_no_client_country`, `test_no_client_postal_code` & `test_no_client_country_postal_code`) could be parameterized and should have a log check. 

```
@pytest.mark.parametrize(
    "geolocation",
    [
        Location(postal_code=94105),
        Location(country="US", region="CA", city="Some City", dma=555),
        Location(),
    ],
)
@pytest.mark.asyncio
async def test_no_client_country_or_postal_code(
    caplog: LogCaptureFixture, accuweather: Provider, geolocation: Location
) -> None:
    """
    Test that if a client has an unknown country or postal code, that a warning is
    logged and no suggestions are returned.
    """
    set_response_bodies()

    res = await accuweather.query(SuggestionRequest("", geolocation))

    assert res == []
    assert len(caplog.messages) == 1
    assert caplog.messages[0] == "Country and/or postal code unknown"
```

**Review by Trinaa — COMMENTED** [review]
Thank you for this work!
From a test perspective it's powerful to be using injectable objects like this!

File: merino/middleware/logging.py (hunk 1 above)
Comment (pjenvey) [reply merino/middleware/logging.py]: Great idea, added a ScopeKey enum

File: merino/providers/base.py [hunk 3]
```diff
@@ -33,11 +44,11 @@ async def initialize(self) -> None:
         ...
 
     @abstractmethod
-    async def query(self, query: str) -> list[BaseSuggestion]:
+    async def query(self, srequest: SuggestionRequest) -> list[BaseSuggestion]:
         """Query against this provider.
 
         Args:
-          - `query`: the query string.
+          - `SuggestionRequest`: the query requested.
```
Comment (ncloudioj) [inline merino/providers/base.py:51]:
```suggestion
          - `SuggestionRequest`: the suggestion request object.
```

Since `geolocation` is also part of `SuggestionRequest`.

**Review by ncloudioj — APPROVED** [review]
Looks great, thanks!

I've run some quick local benchmarks, this runs on par with the `ContextVar` version, but this definitely comes with a couple of benefits over `ContextVar`. So yeah, let's use `scope` to store states onwards 👍

File: merino/middleware/geolocation.py [hunk 4]
```diff
@@ -64,7 +60,7 @@ async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
         except AddressNotFoundError:
             pass
 
-        ctxvar_geolocation.set(
+        scope[ScopeKey.GEOLOCATION] = (
```
Comment (hackebrot) [inline merino/middleware/geolocation.py:63]: I like the idea of using an enum for this to avoid naming conflicts. 👍🏻 

This uses the enum member as the mapping key. I want to make sure that's what we want.

From what I can tell starlette's `Scope` expects `typing.MutableMapping[str, typing.Any]`. This is satisfied for `ScopeKey` since we use `str` as one of the base classes, but it may not be what we want. If want to use `merino_geolocation` as a key in scope, we need to retrieve the value first:

```suggestion
        scope[ScopeKey.GEOLOCATION.value] = (
```

**Review by Trinaa — APPROVED** [review]

File: merino/middleware/geolocation.py (hunk 4 above)
Comment (pjenvey) [reply merino/middleware/geolocation.py:63]: With it being a str subclass I'll argue using the enum variant itself doesn't matter much here as it all works just the same (maybe except you get an enum repr of these keys). Even `scope["merino_geolocation"]` works against the `Enum`.

However if you'd like I can just make `ScopeKey` a regular class of plain str constants instead of an enum.

Comment (hackebrot) [reply merino/middleware/geolocation.py:63]: I think using an `Enum` is the right call here. I just wanted to make sure we're OK with using the enum member for the key rather than the string value.

I didn't know that `scope["merino_geolocation"]` works with that. Let's use an enum!

Comment (pjenvey) [reply merino/middleware/geolocation.py:63]: 👍, reverted back to the enum

File: tests/unit/web/util.py [hunk 5]
```diff
@@ -146,3 +147,8 @@ async def get_providers() -> tuple[dict[str, BaseProvider], list[BaseProvider]]:
 def filter_caplog(records: list[LogRecord], logger_name: str) -> list[LogRecord]:
     """Filter pytest captured log records for a given logger name."""
     return [record for record in records if record.name == logger_name]
+
+
+def srequest(q: str) -> SuggestionRequest:
```
Comment (hackebrot) [inline tests/unit/web/util.py:152]: Long-term, I'd recommend replacing these utility functions with pytest fixtures.

**Review by hackebrot — APPROVED** [review]
LGTM! 🚀
